File: mobile/src/main/java/rhedox/gesahuvertretungsplan/ui/fragment/DatePickerFragment.java
package rhedox.gesahuvertretungsplan.ui.fragment;

import android.app.DatePickerDialog;
import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.widget.DatePicker;

import org.joda.time.DateTime;
import org.joda.time.LocalDate;

import rhedox.gesahuvertretungsplan.ui.activity.MainActivity;

public class DatePickerFragment extends DialogFragment implements DatePickerDialog.OnDateSetListener{
    private LocalDate date;
	private MainActivity mainActivity;
	private boolean isPickerDone = false;

    public static final String TAG ="DatePickerFragment";
    public static final String KEY_DATE ="date";
    public static final String ARGUMENT_DATE = "ArgumentDate";

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState) {
	    if(savedInstanceState != null && savedInstanceState.containsKey(KEY_DATE))
		    date = new DateTime(savedInstanceState.getLong(KEY_DATE)).toLocalDate();
	    else if(getArguments() != null && getArguments().containsKey(ARGUMENT_DATE))
            date = new DateTime(getArguments().getLong(ARGUMENT_DATE)).toLocalDate();
        else
            date = LocalDate.now();

	    isPickerDone = false;
        return new DatePickerDialog(getActivity(), this, date.getYear(), date.getMonthOfYear()-1, date.getDayOfMonth());
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

	    if(context instanceof MainActivity)
		    mainActivity = (MainActivity)context;
    }

    @Override
    public void onDetach() {
        super.onDetach();

	    mainActivity = null;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
	    if(date != null)
            outState.putLong(KEY_DATE, date.toDateTimeAtStartOfDay().getMillis());
    }

	@Override
	public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
		if(!isPickerDone) {
			date = new LocalDate(year, month + 1, dayOfMonth);

			if(mainActivity != null)
				mainActivity.showDate(date);

			isPickerDone = true;
		}

		dismiss();
	}

	public static DatePickerFragment newInstance(LocalDate date)
	{
		Bundle bundle = new Bundle();
		bundle.putLong(ARGUMENT_DATE, date.toDateTimeAtCurrentTime().getMillis());

		DatePickerFragment fragment = new DatePickerFragment();
		fragment.setArguments(bundle);

		return fragment;
	}
}

File: mobile/src/main/java/rhedox/gesahuvertretungsplan/ui/fragment/DatePickerFragment.kt
package rhedox.gesahuvertretungsplan.ui.fragment

import android.app.DatePickerDialog
import android.app.Dialog
import android.os.Bundle
import android.widget.DatePicker
import androidx.fragment.app.DialogFragment
import org.joda.time.DateTime
import org.joda.time.LocalDate
import rhedox.gesahuvertretungsplan.R

/**
 * Created by robin on 20.10.2016.
 */
class DatePickerFragment : DialogFragment(), DatePickerDialog.OnDateSetListener {
    private lateinit var date: LocalDate
    private var isPickerDone = false

    var callback: ((date: LocalDate) -> Unit)? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        date = if (savedInstanceState != null && savedInstanceState.containsKey(keyDate))
            DateTime(savedInstanceState.getLong(keyDate)).toLocalDate()
        else if (arguments != null && arguments!!.containsKey(argumentDate))
            DateTime(arguments!!.getLong(argumentDate)).toLocalDate()
        else
            LocalDate.now()

        isPickerDone = false

        val dialog = DatePickerDialog(requireContext(), this, date.year, date.monthOfYear - 1, date.dayOfMonth)
        dialog.window!!.setBackgroundDrawableResource(R.color.cardBackground) //Workaround for broken/ugly picker look with new theme
        return dialog
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putLong(keyDate, date.toDateTimeAtStartOfDay().millis)
    }

    override fun onDateSet(view: DatePicker, year: Int, month: Int, dayOfMonth: Int) {
        if (!isPickerDone) {
            this.callback?.invoke(LocalDate(year, month+1, dayOfMonth))

            isPickerDone = true
        }

        dismiss()
    }

    companion object {
        const val tag = "DatePickerFragment1"
        const val keyDate = "date"
        const val argumentDate = "ArgumentDate"

        @JvmStatic
        fun newInstance(date: LocalDate): DatePickerFragment {
            val bundle = Bundle()
            bundle.putLong(argumentDate, date.toDateTimeAtCurrentTime().millis)

            val fragment = DatePickerFragment()
            fragment.arguments = bundle

            return fragment
        }
    }
}